cylc/uiserver/resolvers.py

```diff
@@ -349,29 +349,37 @@ async def cat_log(cls, workflow: Tokens, info, task=None):
         # subprocess ends
         asyncio.create_task(cls.enqueue(proc.stdout, queue))
         op_id = info.root_value
+
         while True:
-            if op_id not in info.context['sub_statuses']:
-                break
-            if info.context['sub_statuses'][op_id] == 'stop':
+            # TODO detect if subscription has stopped?
+            if (
+                op_id in info.context['sub_statuses']
+                and info.context['sub_statuses'][op_id] == 'stop'
+            ):
+                print('$$$ stop')
                 with suppress(psutil.Error):
                     cat_log_proc = psutil.Process(proc.pid)
                     # cat-log process will terminate on tail termination
                     for tail_proc in cat_log_proc.children():
                         tail_proc.terminate()
                 # clean the subscription from the status dict
                 info.context['sub_statuses'].pop(op_id)
+                break
 
-            if queue.empty() and buffer:
```

Comment:
If both the queue and the buffer were empty this would skip to the `else` branch where it would `await queue.get`. I think this may have been causing a deadlock, dunno, split the branch to prevent it happening.